App installation auth

rocallahan opened the pull request:
This implements what I had proposed in issue #1. Sorry about being slow to publish it, I only recently got our app working well (https://github.com/Pernosco/pernosco/wiki/Using-Pernosco-To-Debug-Github-Actions-Test-Failures).

I haven't touched the `create_jwt` that was recently added. We could leave it as-is, or we could consolidate the implementations, or delete it.

rocallahan commented:
Also if you ever want to debug CI test failures then we could enable Pernosco for your repo.

XAMPPRocky commented on src/lib.rs:
```diff
@@ -664,9 +766,73 @@ impl Octocrab {
+        };
+        let mut retries = 0;
+        loop {
+            let result = self.client.post(self.absolute_url(format!("/app/installations/{}/access_tokens", installation))?)
```
Shouldn't there be a `access_tokens` method here in the octocrab API that you can use instead of calling it directly?

rocallahan replied:
There could be, but as far as I can tell, this is the only reasonable way to use access tokens. You want them to auto-refresh, and you're going to use them to authenticate other Github requests, so why would you use a stand-alone `access_tokens` API call?

XAMPPRocky replied:
Well someone wants to implement their own authentication for whatever reason, having this would make that slightly easier.

XAMPPRocky commented on src/lib.rs:
```diff
@@ -497,6 +597,8 @@ impl Octocrab {
     }
 }
 
+const MAX_RETRIES: u32 = 3;
```
Can we move this to the very top of the file?

rocallahan replied:
Sure

XAMPPRocky commented on src/lib.rs:
```diff
@@ -329,15 +343,76 @@ impl OctocrabBuilder {
 }
 
+/// A cached API access token (which may be None)
+struct CachedToken(Mutex<Option<String>>);
```
This should be a `RwLock`, as you're going to be reading this more than writing to it.

rocallahan replied:
Sure

XAMPPRocky reviewed (changes requested):
Thank you for your PR! This mostly looks good to me. I think we should consolidate the implementations to make sure it's consistent.

rocallahan commented:
> I think we should consolidate the implementations to make sure it's consistent.
 
OK, the only problem is that my code accepts a key in DER format and `create_jwt` accepts a key in PEM format. That's annoying. I think probably to accommodate both use-cases, `OctocrabBuilder::app()` should have its key parameter be `jsonwebtoken::EncodingKey` so the caller can choose any format supported by `jsonwebtoken`. `create_jwt` can remain unchanged (only accepts PEM format) for now.

rocallahan commented:
I'm going to push the updated branch, but there is a problem: I don't have any tests for this stuff other than my actual live testing on Github with Pernosco. So I don't recommend landing changes until I've incorporated these changes downstream. Do you have any suggestions here?

rocallahan commented:
I've also noticed that some of the existing APIs and models don't actually work, probably for similar reasons. I'll push PRs for those fixes later.

XAMPPRocky commented:
Hmm, I think we should remove accepting PEM, PEM is just base64 DER, so if you need to decode it first, you can always use the `pem` crate to decode it first.

Yeah, I don't have a great solution for that at the moment. My hope is eventually to remove a bunch of the boilerplate, and then autogenerate the definitions, but I haven't found the time.

rocallahan commented:
> Hmm, I think we should remove accepting PEM, PEM is just base64 DER, so if you need to decode it first, you can always use the pem crate to decode it first.

Ok

rocallahan commented:
I just changed `create_jwt` to take a `jsonwebtoken::EncodingKey` so people can pass in whatever they want. [It's trivial](https://docs.rs/jsonwebtoken/7.2.0/jsonwebtoken/struct.EncodingKey.html) to use tat `EncodingKey` API to consume DER or PEM.

rocallahan commented:
That should fix the build failure. I also moved the `github_app_authentication.rs` to `github_app_authentication_manual.rs` and replaced `github_app_authentication.rs` with an example using the new `OctocrabBuilder::app()` API, which is a lot simpler.
